clang/lib/Sema/BoundsAnalysis.cpp

```diff
@@ -136,128 +133,174 @@ bool BoundsAnalysis::AreDeclaredBoundsZero(const Expr *E, const Expr *V) {
+    // bounds are bounds(p, p) or count(0). We need to generalize this to
+    // widen bounds for dereferences involving constant offsets from the
+    // declared upper bound of a variable.
+    if (!AreDeclaredBoundsZero(AE->getBoundsExpr(), D))
+      return;
 
-      Expr *LHS = IgnoreCasts(BO->getLHS());
-      Expr *RHS = IgnoreCasts(BO->getRHS());
-      DeclRefExpr *D = nullptr;
-      IntegerLiteral *Lit = nullptr;
-
-      // Handle *(p + i).
-      if (isa<DeclRefExpr>(LHS) && isa<IntegerLiteral>(RHS)) {
-        D = dyn_cast<DeclRefExpr>(LHS);
-        Lit = dyn_cast<IntegerLiteral>(RHS);
-
-      // Handle *(i + p).
-      } else if (isa<DeclRefExpr>(RHS) && isa<IntegerLiteral>(LHS)) {
-        D = dyn_cast<DeclRefExpr>(RHS);
-        Lit = dyn_cast<IntegerLiteral>(LHS);
-      }
+    if (const auto *V = dyn_cast<VarDecl>(D->getDecl())) {
+      if (IsNtArrayType(V)) {
+        Expr::EvalResult Res;
+        if (!Index->EvaluateAsInt(Res, S.Context))
+          return;
 
-      if (!D || !Lit)
-        return;
+        llvm::APSInt Idx = Res.Val.getInt();
+        if (Idx.isNegative())
+          return;
 
-      // TODO: Remove this check. Currently, for the first version of this
-      // algorithm, we are enabling bounds widening only when the declared
-      // bounds are bounds(p, p) or count(0). We need to generalize this to
-      // widen bounds for dereferences involving constant offsets from the
-      // declared upper bound of a variable.
-      if (!AreDeclaredBoundsZero(UO->getBoundsExpr(), D))
-        return;
+        EB->Gen[SuccEB->Block].insert(std::make_pair(V, Idx.getLimitedValue()));
 
-      if (const auto *V = dyn_cast<VarDecl>(D->getDecl())) {
-        if (V->getType()->isCheckedPointerNtArrayType()) {
-          // We update the bounds of p on the edge EB->SuccEB only if this is
-          // the first time we encounter "if (*(p + i)" on that edge.
-          if (!EB->Gen[SuccEB->Block].count(V)) {
-            EB->Gen[SuccEB->Block][V] = Lit->getValue().getLimitedValue();
-            if (!SuccEB->BoundsVars.count(V)) {
-              DeclSetTy BoundsVars;
-              CollectBoundsVars(UO->getBoundsExpr(), BoundsVars);
-              SuccEB->BoundsVars[V] = BoundsVars;
-            }
-          }
+        if (!SuccEB->BoundsVars.count(V)) {
+          DeclSetTy BoundsVars;
+          CollectBoundsVars(AE->getBoundsExpr(), BoundsVars);
+          SuccEB->BoundsVars[V] = BoundsVars;
         }
       }
     }
   }
 }
 
+void BoundsAnalysis::FillGenSet(Expr *E,
+                                ElevatedCFGBlock *EB,
+                                ElevatedCFGBlock *SuccEB) {
+
+  // Handle if conditions of the form "if (*e1 && *e2)".
+  if (const auto *BO = dyn_cast<const BinaryOperator>(E)) {
+    if (BO->getOpcode() == BO_LAnd) {
+      FillGenSet(BO->getLHS(), EB, SuccEB);
+      FillGenSet(BO->getRHS(), EB, SuccEB);
+    }
+  }
+
+  E = IgnoreCasts(E);
```

Comment:
I think there should be an explicit check for an lvalue-to-rvalue cast here, instead of doing it as part of IgnoreCasts.